### ISSUE TITLE
[2.0a0 AutoGraph] segmentation fault after refactoring into smaller functions

### ISSUE THREAD

**Issue by llan-ml** [tensorflow/tensorflow#29488]
**System information**
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0alpha0
- Python version: 3.6.5

**Code to reproduce the issue**
<pre>
# -*- coding: utf-8 -*-
# @Author  : Lin Lan (ryan.linlan@gmail.com)

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np
import tensorflow as tf

class Model(tf.keras.Model):
    def __init__(self):
        super(Model, self).__init__()
        self.dense = tf.keras.layers.Dense(10)

    def call(self, inputs):
        return self.dense(inputs)

model = Model()

def forward(x):
    batch_size = x.shape[0]
    ys = tf.TensorArray(tf.float32, size=batch_size)
    for i in tf.range(batch_size):
        y = model(x[i][tf.newaxis, :])
        ys = ys.write(i, y)
    return ys.stack()

def train(x, forward_func):
    with tf.GradientTape() as tape:
        ys = forward_func(x)
        loss = tf.reduce_mean(ys)
    grads = tape.gradient(loss, model.trainable_weights)
    return grads

def big_train(x, forward_func):
    with tf.GradientTape() as tape:
        batch_size = x.shape[0]
        ys = tf.TensorArray(tf.float32, size=batch_size)
        for i in tf.range(batch_size):
            y = model(x[i][tf.newaxis, :])
            ys = ys.write(i, y)
        ys = ys.stack()
        loss = tf.reduce_mean(ys)
    grads = tape.gradient(loss, model.trainable_weights)
    return grads

x = np.random.rand(10, 100).astype(np.float32)

train(x, forward)
print("pass")

tf.function(big_train)(x, tf.function(forward))
print("pass")

tf.function(train)(x, tf.function(forward))    # lead to segmentation fault (core dumpled)
print("pass")
</pre>

**Other info / logs**
Possibly related to #29393 . Also, due to that issue, we have to decorate some nested methods with `tf.function`.
@alextp @mdanatg

**Comment by llan-ml** [comment]
@martinwicke I ran the above code with `tf-nightly-gpu-2.0-preview 2.0.0.dev20190606`, and the segmentation fault disappeared. However, there are some other errors. Thus, I prefer to close this issue and open another.
